### DISCUSSION TITLE
Note Ontologies

### DISCUSSION THREAD

**Discussion by TheCedarPrince** [TheCedarPrince/NoteMate.jl#25] [General]
Thoughts from @SevorisDoe in issue #21: 

> @TheCedarPrince: Hey, whenever you have the time/desire @SevorisDoe , could you explain to me more about what you mean by an ontology in the discussions section? I am not fully grasping what you mean here but I want to as I think it is really powerful/useful vocabulary. Thanks!

So for me, knowledge graphs have framed the term ontology. The notion here is that an ontology is a definition of concepts, entities and their relations (on which you can compute). You could also call it a logical specification. I think this definition describs it well:

> An ontology is a formal, explicit specification of a shared conceptualization that is characterized by high semantic expressiveness required for increased complexity.
 > ― Feilmayr, Christina; Wöß, Wolfram (2016). "An analysis of ontologies and their success factors for application to business". Data & Knowledge Engineering. 101: 1–23. DOI[10.1016/j.datak.2015.11.003](https://doi.org/10.1016%2Fj.datak.2015.11.003).

In this case an ontology would be the definition of what things you find in a note, and what notes you find in a vault, what structures they have, and how those relate - i.e. OKM is a notes ontology because we can define that the bullets in the references section (defined by a heading with a certain name, until the subsequent header) each refer to some other thing. Or the three bold-starting paragraphs under the title (which is the first H1 header of the document) are the date, the summary, and the keywords, with each keyword being a hash separated by a whitespace...

In my vault for example, notes in the "50 Ressources/Zotero" folder are notes whose title matches Bibtex keys in my Zotero library. Notes that sit in my "Projects" folder are project notes, and so for example "Projects/NoteMate.md" is a project note about this effort here, and any note that links to it refers to the project. This kind of information is encoded in the ontology.

The point for us here is basically, an ontology defines labeled entitens within notes, and to the notes themselves. It creates entities which we can define to have certain properties, and which we can then target by their relations. I.e. we might use the ontology to say "any note which has resides in the Zettelkasten folder, and has a literary tag, is a literary note and thus we consider it a reference. We know that literary notes have a header where the third paragraph starting with the string "link" specifies the URL and that below that we find a "source" text, which we can copy to create the references later.

Those are specific examples. The general view is more unconstrained, but the consequence is a kind of "structured labelling" which you can use for computation and manipulation of notes by undestanding what entities are defined in them, what the note itself represents within the larger workspace, and how you should interpret links between them (see also the keyword of semantic links)

**Comment by TheCedarPrince** [comment]
> So for me, knowledge graphs have framed the term ontology. The notion here is that an ontology is a definition of concepts, entities and their relations (on which you can compute). You could also call it a logical specification.

Thanks for sharing your understanding on this! For me, in health informatics research, we have a similar yet different meaning. Diseases, symptoms, procedures, etc. fall under a clinical ontology that can be used to label a particular group of symptoms and conditions using a specific vocabulary. We also refer to this entire process as phenotyping which can also introduce language confusion. 

> > An ontology is a formal, explicit specification of a shared conceptualization that is characterized by high semantic expressiveness required for increased complexity.
> > ― Feilmayr, Christina; Wöß, Wolfram (2016). "An analysis of ontologies and their success factors for application to business". Data & Knowledge Engineering. 101: 1–23. DOI[10.1016/j.datak.2015.11.003](https://doi.org/10.1016%2Fj.datak.2015.11.003).

I like this definition a lot; thanks for sharing it.

> The point for us here is basically, an ontology defines labeled entitens within notes, and to the notes themselves.

This, alongside your examples, really helped everything click into place for my understanding. So, like with the Targets I defined so far (namely just the `FranklinNote`), these would be their own ontologies: like an ontology for Franklin Notes, an ontology for Obsidian Notes, etc. And the OKM itself is, if I follow you correctly, an ontology for Notes in general -- which would make sense to me as that is what I was trying to do. 

In terms of turning this discussion into something more actionable, what is your opinion on the documentation having a section discussing the different sort of note ontologies supported? It kinda would support this issue: #22 and enriches the utility of the API page. Thoughts?

**Comment by SevorisDoe** [comment]
The big challenge is that while ontologies are evidently something of practical concern, they are also very theoretical, and we started on the far side of the matter. NoteMate has no templating language and compiler, nor an ontology processor that can match specifications to inputs, nor a processing and templating that can use the labeling. So we have two very disparate issues, and I am somewhat hesitant to take the term into my mouth for concrete, present-term implementations at right now, even if the entire theory is incredibly interesting - and if executed in well, would have far-reaching consequences. It is also IMO the kind of thing you kind of need to really, really kick something like Julia forward as a language for working on one's vaults. 

If we want to work towards actionables, we have to think about a couple of things. 
- how do we frame the present question of document layouts, what information we require, and how we get it? 
- how do we extend or redesign the current API?
- how do we frame working with documents
- how do we frame our extension of support? Towards what features, using what infrastructure, do we extend coverage? How does the practical short-term vision fit together with the long-term vision? 

There are a couple of tricky issues here. We might want to brainstorm about this sometimes. I find it difficult to fully describe the problem and implication space right now.
